File: perl/immersunproxy2mqtt.pl
#!/usr/bin/perl
#
# Copyright (c) 2020 David M Brooke, davidmbrooke@users.sourceforge.net
#
#
use strict;
use warnings;
use DateTime;
use AnyEvent::MQTT;
use AnyEvent::Handle::UDP;	# libanyevent-handle-udp-perl 

# File-scope variables
our $mqtt;
our $logfile;
our $udp_server;
our $udp_client;
our $ob_data;
our $ob_ae_handle;
our $ob_client_addr;

# Subroutine for Outbound Callback
sub ob_rx_callback {
    ( $ob_data, $ob_ae_handle, $ob_client_addr ) = @_;
    my $cv_mqtt;
    my $bytestring;

    # Generate date/timestamp for logfile
    my $datetime = DateTime->now( )->iso8601( ).'Z';

    # The "exporting" value is a signed Short at byte 29 & 30
    my $exporting = unpack( "x28 s<", $ob_data );
    # The "generating" value is a signed Short at byte 53 & 54
    my $generating = unpack( "x52 s<", $ob_data );
    # The "diverting" value is a signed Short at byte 55 & 56
    my $diverting = unpack( "x54 s<", $ob_data );
    # The "consuming" value is calculated from the others
    my $consuming = 0 - ( $exporting - $generating + $diverting );
    my $importing = 0 - $exporting;

    # Convert data to hex for logfile
    $bytestring = unpack( 'H*', $ob_data );
    ##print $logfile "$datetime $bytestring\n";
    ##print "$datetime  Exporting: $exporting Generating: $generating Diverting: $diverting Consuming: $consuming\n";

    # Check for 'odd' datagram format - must be a better test than this but need to investigate further
    if ( $generating != 0 )
    {
        # Publish the data to MQTT
        $cv_mqtt = $mqtt->publish( topic => "raw/immersun/220172/generating/power", message => $generating );
        $cv_mqtt = $mqtt->publish( topic => "raw/immersun/220172/exporting/power", message => $exporting );
        $cv_mqtt = $mqtt->publish( topic => "raw/immersun/220172/importing/power", message => $importing );
        $cv_mqtt = $mqtt->publish( topic => "raw/immersun/220172/diverting/power", message => $diverting );
        $cv_mqtt = $mqtt->publish( topic => "raw/immersun/220172/consuming/power", message => $consuming );
    }

    # Forward the Outbound packet to live.immersun.com
    my $cv = $udp_client->push_send( $ob_data );
}

# Subroutine for Inbound Callback
sub ib_rx_callback {
    my ( $data, $ae_handle, $client_addr ) = @_;

    # Forward the Inbound packet to the immerLINK Bridge
    my $cv = $udp_server->push_send( $data, $ob_client_addr );
}


# Main Program

# Open the logfile
#open( $logfile, ">>", "/home/pi/immersun.log" )
    #or die "Can't open output file: $!";
#$logfile->autoflush( 1 );


# Connect to the MQTT Broker
$mqtt = AnyEvent::MQTT->new( host => 'mqtt',
			     client_id => 'immersun2mqtt',
			     user_name => 'immersun',
			     password => '<secret>' );
die unless defined $mqtt;

# Create the UDP Server
$udp_server = AnyEvent::Handle::UDP->new( bind => ['172.16.40.81', 87],
					  on_recv => \&ob_rx_callback );
die unless defined $udp_server;

# Create the UDP Client
$udp_client = AnyEvent::Handle::UDP->new( connect => ['136.243.233.46', 87],
					  on_recv => \&ib_rx_callback );
die unless defined $udp_client;

# Run the event loop
AnyEvent->condvar->recv;

# Shouldn't ever get here but just in case...
#close( $logfile );

File: python/immerSUN-proxy2mqtt.py
"""immerSUN UDP proxy server"""
#
#   Adapted from https://gist.github.com/vxgmichel/b2cf8536363275e735c231caef35a5df
#   (copied at 2022-12-04T17:00)
#
#   This is a UDP proxy server for an immerSUN solar generation diverter with optional
#   immerLINK bridge that is sending readings to myimmerSUN.com.
#
#   If UDP datagrams are artifically directed at this proxy server (rather than being
#   sent direct to the myimmerSUN.com server) they will have various data values extracted 
#   and then the original datagrams will be forwarded unmodified.
#   The response (inbound) datagrams will be forwarded back to the immerLINK bridge.
#   The extracted data values will be published as MQTT messages.
#
#
#   Configuration settings for the UDP Proxy Server:
#       The BIND_ADDRESS is the IPv4 address this proxy should listen on
#           0.0.0.0 will listen on all local addresses
#           If you have multiple local addresses and you want to use a particular one,
#           specify it here
BIND_ADDRESS = "0.0.0.0"
#       The REMOTE_HOST is the IPv4 address this proxy should forward datagrams to
#           136.243.233.46 is the 'Cloud IP' for the myimmerSUN.com server (as of 2022-12)
REMOTE_HOST = "136.243.233.46"
#       The UDP_PORT
UDP_PORT = 87
#
#   Configuration settings for MQTT:
#       The hostname or IPv4 address of the MQTT Broker
MQTT_BROKER = "mymqttbroker"
#       The MQTT port number 
MQTT_PORT = 1883
#       The MQTT username
MQTT_USER = "immersunuser"
#       The MQTT password
MQTT_PASSWORD = "mysecretpassword"
#       The MQTT client id
MQTT_CLIENT = "immersun2mqtt"
#       The MQTT Topic base path
MQTT_BASE_TOPIC = "raw/immersun/"
#
#
DEBUG = False


import asyncio
import paho.mqtt.client as mqtt


class ProxyDatagramProtocol(asyncio.DatagramProtocol):

    def __init__(self, remote_address, mqtt_client):
        self.remote_address = remote_address
        self.mqtt_client = mqtt_client
        self.remotes = {}
        super().__init__()

    def connection_made(self, transport):
        self.transport = transport

    def datagram_received(self, data, addr):
        # This is the outbound *request* to the immerSUN server
        # 'data' is a 'bytes object' - generally 56 bytes long
        numbytes = len(data)
        if (DEBUG):
            print("Outbound datagram... " + str(numbytes) + " bytes long")
            print(data.hex())
        # Extract the known data fields, stored in little-endian byte order
        exporting  = int.from_bytes(data[28:30], "little", signed=True)
        serial     = int.from_bytes(data[36:40], "little", signed=True)
        generating = int.from_bytes(data[52:54], "little", signed=True)
        diverting  = int.from_bytes(data[54:56], "little", signed=True)
        consuming = 0 - (exporting - generating + diverting)
        importing = 0 - exporting
        #
        #   Check for '3094' condition
        if exporting == 3094 and consuming == -3094:
            print("3094 condition detected - data values ignored")
        else:
            if (DEBUG):
                print("Exporting:  " + str(exporting))
                print("Generating: " + str(generating))
                print("Diverting:  " + str(diverting))
                print("Consuming:  " + str(consuming))
            #
            # Publish the gathered data to MQTT
            topic = MQTT_BASE_TOPIC + str(serial) + "/generating/power"
            result = self.mqtt_client.publish(topic, generating)
            if result.rc != 0:
                print("Error publishing to MQTT Broker: " + str(result.rc))
            topic = MQTT_BASE_TOPIC + str(serial) + "/consuming/power"
            result = self.mqtt_client.publish(topic, consuming)
            if result.rc != 0:
                print("Error publishing to MQTT Broker: " + str(result.rc))
            topic = MQTT_BASE_TOPIC + str(serial) + "/diverting/power"
            result = self.mqtt_client.publish(topic, diverting)
            if result.rc != 0:
                print("Error publishing to MQTT Broker: " + str(result.rc))
            topic = MQTT_BASE_TOPIC + str(serial) + "/importing/power"
            result = self.mqtt_client.publish(topic, importing)
            if result.rc != 0:
                print("Error publishing to MQTT Broker: " + str(result.rc))
        #
        if addr in self.remotes:
            self.remotes[addr].transport.sendto(data)
            return
        loop = asyncio.get_event_loop()
        self.remotes[addr] = RemoteDatagramProtocol(self, addr, data)
        coro = loop.create_datagram_endpoint(
            lambda: self.remotes[addr], remote_addr=self.remote_address)
        asyncio.ensure_future(coro)


class RemoteDatagramProtocol(asyncio.DatagramProtocol):

    def __init__(self, proxy, addr, data):
        self.proxy = proxy
        self.addr = addr
        self.data = data
        super().__init__()

    def connection_made(self, transport):
        self.transport = transport
        self.transport.sendto(self.data)

    def datagram_received(self, data, _):
        # This is the inbound *response* from the immerSUN server
        # 'data' is a 'bytes object' - expected to always be 56 bytes long
        numbytes = len(data)
        if (DEBUG):
            print("Inbound datagram...  " + str(numbytes) + " bytes long")
            print(data.hex())
        self.proxy.transport.sendto(data, self.addr)

    def connection_lost(self, exc):
        self.proxy.remotes.pop(self.attr)


async def start_datagram_proxy(bind, port, remote_host, remote_port, mqtt_client):
    loop = asyncio.get_event_loop()
    protocol = ProxyDatagramProtocol((remote_host, remote_port), mqtt_client)
    # create_datagram_endpoint() returns an instance of class 'DatagramTransport'
    return await loop.create_datagram_endpoint(
        lambda: protocol, local_addr=(bind, port))


#
# This is the 'main' function that starts everything up then loops forever
#
def main(bind=BIND_ADDRESS, port=UDP_PORT, remote_host=REMOTE_HOST, remote_port=UDP_PORT):
    loop = asyncio.get_event_loop()

    # Open the connection to the MQTT server
    if (DEBUG):
        print("Connecting to MQTT server...")
    mqtt_client = mqtt.Client(MQTT_CLIENT)
    mqtt_client.username_pw_set(MQTT_USER, MQTT_PASSWORD)
    mqtt_client.connect(MQTT_BROKER, MQTT_PORT)

    # Start the UDP proxy
    if (DEBUG):
        print("Starting datagram proxy...")
    coro = start_datagram_proxy(bind, port, remote_host, remote_port, mqtt_client)
    transport, _ = loop.run_until_complete(coro)
    if (DEBUG):
        print("Datagram proxy is running...")
    try:
        loop.run_forever()
    except KeyboardInterrupt:
        pass
    if (DEBUG):
        print("Disconnecting from MQTT server...")
    mqtt_client.disconnect()
    if (DEBUG):
        print("Closing transport...")
    transport.close()
    loop.close()


if __name__ == '__main__':
    main()

